### PR TITLE
Create new directory only if not already present

### PR DESCRIPTION
**Pull request by Harshg999**
## What changes were proposed in this pull request?

- Create a new directory only when it is not already present in the parent directory to avoid duplication
- Also manually deleted all the duplicated Gist directories except one

### CODE REVIEW

File: desktop/core/src/desktop/api2.py [hunk 1]
```diff
@@ -520,13 +520,23 @@ def create_directory(request):
 
   parent_dir = Directory.objects.get_by_uuid(user=request.user, uuid=parent_uuid, perm_type='write')
 
-  directory = Directory.objects.create(name=name, owner=request.user, parent_directory=parent_dir)
-
-  return JsonResponse({
+  # Check if directory already exists then return it to avoid creating duplicate directory
+  try:
+    dir_exist = Directory.objects.get(name=name, owner=request.user, parent_directory=parent_dir)
```
Comment (romainr) [inline desktop/core/src/desktop/api2.py:525]: Use insted https://docs.djangoproject.com/en/3.1/ref/models/querysets/#get-or-create ?

`dir, created = get_or_create()`

**Review by romainr — COMMENTED** [review]
Nice!

Just I am not sure that you are fixing the actual issue. Did you create a test to show the repro first?

PS: check those lines for inspiration:
```
get_gist_directory
get_home_directory
test_gist_directory_creation
GIST_DIR = 'Gist'

```

And I just dumped the gist dirs from the demo. It shows why it failed:
```

n [4]: from desktop.models import Directory, Document2

In [5]: [d.to_dict() for d in Directory.objects.filter(name=Document2.GIST_DIR)]
Out[5]: 
[{'owner': 'demo',
  'name': 'Gist',
  'path': '/.Trash/Gist',
  'description': '',
  'uuid': '3b1dc1ec-01a3-4fa6-906d-2330e893f9f9',
  'id': 98306,
  'doc1_id': 94314,
  'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
  'type': 'directory',
  'perms': {'read': {'users': [], 'groups': []},
   'write': {'users': [], 'groups': []},
   'link_sharing_on': False,
   'link_read': False,
   'link_write': False},
  'last_modified': '2021-02-28T13:38Z',
  'last_modified_ts': 1614519484,
  'is_managed': False,
  'isSelected': False,
  'absoluteUrl': '/home2?uuid=3b1dc1ec-01a3-4fa6-906d-2330e893f9f9'},
 {'owner': 'demo',
  'name': 'Gist',
  'path': '/.Trash/Gist',
  'description': '',
  'uuid': '3b8d29fd-7b26-4d3e-a018-3aff68331bf2',
  'id': 101543,
  'doc1_id': 97429,
  'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
  'type': 'directory',
  'perms': {'read': {'users': [], 'groups': []},
   'write': {'users': [], 'groups': []},
   'link_sharing_on': False,
   'link_read': False,
   'link_write': False},
  'last_modified': '2021-02-28T13:37Z',
  'last_modified_ts': 1614519472,
  'is_managed': False,
  'isSelected': False,
  'absoluteUrl': '/home2?uuid=3b8d29fd-7b26-4d3e-a018-3aff68331bf2'},
 {'owner': 'demo',
  'name': 'Gist',
  'path': '/.Trash/Gist',
  'description': '',
  'uuid': '1c113f15-0e21-4ef3-8279-1cd4ef96ecdd',
  'id': 102313,
  'doc1_id': 98169,
  'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
  'type': 'directory',
  'perms': {'read': {'users': [], 'groups': []},
   'write': {'users': [], 'groups': []},
   'link_sharing_on': False,
   'link_read': False,
   'link_write': False},
  'last_modified': '2021-02-28T13:37Z',
  'last_modified_ts': 1614519458,
  'is_managed': False,
  'isSelected': False,
  'absoluteUrl': '/home2?uuid=1c113f15-0e21-4ef3-8279-1cd4ef96ecdd'},
 {'owner': 'demo',
  'name': 'Gist',
  'path': '/.Trash/Gist',
  'description': '',
  'uuid': '6a285c00-ec6e-4ac1-8183-04a8ba818db6',
  'id': 108311,
  'doc1_id': 103954,
  'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
  'type': 'directory',
  'perms': {'read': {'users': [], 'groups': []},
   'write': {'users': [], 'groups': []},
   'link_sharing_on': False,
   'link_read': False,
   'link_write': False},
  'last_modified': '2021-02-28T13:37Z',
  'last_modified_ts': 1614519445,
  'is_managed': False,
  'isSelected': False,
  'absoluteUrl': '/home2?uuid=6a285c00-ec6e-4ac1-8183-04a8ba818db6'},
 {'owner': 'demo',
  'name': 'Gist',
  'path': '/.Trash/Gist',
  'description': '',
  'uuid': '84abbc88-74f8-4b0a-8334-43ad1b45f524',
  'id': 108314,
  'doc1_id': 103968,
  'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
  'type': 'directory',
  'perms': {'read': {'users': [], 'groups': []},
   'write': {'users': [], 'groups': []},
   'link_sharing_on': False,
   'link_read': False,
   'link_write': False},
  'last_modified': '2021-02-28T13:29Z',
  'last_modified_ts': 1614518972,
  'is_managed': False,
  'isSelected': False,
  'absoluteUrl': '/home2?uuid=84abbc88-74f8-4b0a-8334-43ad1b45f524'},
 {'owner': 'demo',
  'name': 'Gist',
  'path': '/.Trash/Gist',
  'description': '',
  'uuid': '9b36c93e-4c0b-42b8-b18a-f3826dbff271',
  'id': 108318,
  'doc1_id': 103971,
  'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
  'type': 'directory',
  'perms': {'read': {'users': [], 'groups': []},
   'write': {'users': [], 'groups': []},
   'link_sharing_on': False,
   'link_read': False,
   'link_write': False},
  'last_modified': '2021-02-28T13:29Z',
  'last_modified_ts': 1614518957,
  'is_managed': False,
  'isSelected': False,
  'absoluteUrl': '/home2?uuid=9b36c93e-4c0b-42b8-b18a-f3826dbff271'},
 {'owner': 'demo',
  'name': 'Gist',
  'path': '/.Trash/Gist',
  'description': '',
  'uuid': 'd24b175d-7877-4693-9e29-a749aa96216c',
  'id': 108321,
  'doc1_id': 103975,
  'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
  'type': 'directory',
  'perms': {'read': {'users': [], 'groups': []},
   'write': {'users': [], 'groups': []},
   'link_sharing_on': False,
   'link_read': False,
   'link_write': False},
  'last_modified': '2021-02-28T13:29Z',
  'last_modified_ts': 1614518944,
  'is_managed': False,
  'isSelected': False,
  'absoluteUrl': '/home2?uuid=d24b175d-7877-4693-9e29-a749aa96216c'},
 {'owner': 'demo',
  'name': 'Gist',
  'path': '/.Trash/Gist',
  'description': '',
  'uuid': '206e33d2-04fd-4b68-8d3d-d5231b1b5192',
  'id': 108326,
  'doc1_id': 103978,
  'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
  'type': 'directory',
  'perms': {'read': {'users': [], 'groups': []},
   'write': {'users': [], 'groups': []},
   'link_sharing_on': False,
   'link_read': False,
   'link_write': False},
  'last_modified': '2021-02-28T13:28Z',
  'last_modified_ts': 1614518915,
  'is_managed': False,
  'isSelected': False,
  'absoluteUrl': '/home2?uuid=206e33d2-04fd-4b68-8d3d-d5231b1b5192'},
 {'owner': 'demo',
  'name': 'Gist',
  'path': '/.Trash/Gist',
  'description': '',
  'uuid': 'b6807cda-9601-45c6-b43f-cb811dc6ca6c',
  'id': 108329,
  'doc1_id': 103983,
  'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
  'type': 'directory',
  'perms': {'read': {'users': [], 'groups': []},
   'write': {'users': [], 'groups': []},
   'link_sharing_on': False,
   'link_read': False,
   'link_write': False},
  'last_modified': '2021-02-28T13:28Z',
  'last_modified_ts': 1614518898,
  'is_managed': False,
  'isSelected': False,
  'absoluteUrl': '/home2?uuid=b6807cda-9601-45c6-b43f-cb811dc6ca6c'},
 {'owner': 'demo',
  'name': 'Gist',
  'path': '/.Trash/Gist',
  'description': '',
  'uuid': '444727f8-5446-44cb-b554-2f759cd9092f',
  'id': 108332,
  'doc1_id': 103986,
  'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
  'type': 'directory',
  'perms': {'read': {'users': [], 'groups': []},
   'write': {'users': [], 'groups': []},
   'link_sharing_on': False,
   'link_read': False,
   'link_write': False},
  'last_modified': '2021-02-28T13:28Z',
  'last_modified_ts': 1614518884,
  'is_managed': False,
  'isSelected': False,
  'absoluteUrl': '/home2?uuid=444727f8-5446-44cb-b554-2f759cd9092f'},
 {'owner': 'demo',
  'name': 'Gist',
  'path': '/.Trash/Gist',
  'description': '',
  'uuid': '4c85cf1f-8857-4f54-bbba-ede6f6ce5a12',
  'id': 108337,
  'doc1_id': 103989,
  'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
  'type': 'directory',
  'perms': {'read': {'users': [], 'groups': []},
   'write': {'users': [], 'groups': []},
   'link_sharing_on': False,
   'link_read': False,
   'link_write': False},
  'last_modified': '2021-02-28T11:34Z',
  'last_modified_ts': 1614512080,
  'is_managed': False,
  'isSelected': False,
  'absoluteUrl': '/home2?uuid=4c85cf1f-8857-4f54-bbba-ede6f6ce5a12'},
 {'owner': 'demo',
  'name': 'Gist',
  'path': '/.Trash/Gist',
  'description': '',
  'uuid': '869eb70a-964a-4677-a6cd-7c1ef45566db',
  'id': 108340,
  'doc1_id': 103994,
  'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
  'type': 'directory',
  'perms': {'read': {'users': [], 'groups': []},
   'write': {'users': [], 'groups': []},
   'link_sharing_on': False,
   'link_read': False,
   'link_write': False},
  'last_modified': '2021-02-28T11:34Z',
  'last_modified_ts': 1614512064,
  'is_managed': False,
  'isSelected': False,
  'absoluteUrl': '/home2?uuid=869eb70a-964a-4677-a6cd-7c1ef45566db'},
 {'owner': 'demo',
  'name': 'Gist',
  'path': '/Gist',
  'description': '',
  'uuid': 'dcec6b13-1317-4a3b-98b1-621ac29542b4',
  'id': 92640,
  'doc1_id': 90136,
  'parent_uuid': 'e214bac4-e46b-4fce-93f6-199c10d89a3c',
  'type': 'directory',
  'perms': {'read': {'users': [], 'groups': []},
   'write': {'users': [], 'groups': []},
   'link_sharing_on': False,
   'link_read': False,
   'link_write': False},
  'last_modified': '2021-01-20T18:02Z',
  'last_modified_ts': 1611165721,
  'is_managed': False,
  'isSelected': False,
  'absoluteUrl': '/home2?uuid=dcec6b13-1317-4a3b-98b1-621ac29542b4'}]

```

File: desktop/core/src/desktop/api2.py (hunk 1 above)
Comment (romainr) [reply desktop/core/src/desktop/api2.py:525]: (actually better to not change it as "probably" not related to problem)

**Comment by Harshg999** [comment]
> Nice!
> 
> Just I am not sure that you are fixing the actual issue. Did you create a test to show the repro first?
> 
> PS: check those lines for inspiration:
> 
> ```
> get_gist_directory
> get_home_directory
> test_gist_directory_creation
> GIST_DIR = 'Gist'
> ```
> 
> And I just dumped the gist dirs from the demo. It shows why it failed:
> 
> ```
> 
> n [4]: from desktop.models import Directory, Document2
> 
> In [5]: [d.to_dict() for d in Directory.objects.filter(name=Document2.GIST_DIR)]
> Out[5]: 
> [{'owner': 'demo',
>   'name': 'Gist',
>   'path': '/.Trash/Gist',
>   'description': '',
>   'uuid': '3b1dc1ec-01a3-4fa6-906d-2330e893f9f9',
>   'id': 98306,
>   'doc1_id': 94314,
>   'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
>   'type': 'directory',
>   'perms': {'read': {'users': [], 'groups': []},
>    'write': {'users': [], 'groups': []},
>    'link_sharing_on': False,
>    'link_read': False,
>    'link_write': False},
>   'last_modified': '2021-02-28T13:38Z',
>   'last_modified_ts': 1614519484,
>   'is_managed': False,
>   'isSelected': False,
>   'absoluteUrl': '/home2?uuid=3b1dc1ec-01a3-4fa6-906d-2330e893f9f9'},
>  {'owner': 'demo',
>   'name': 'Gist',
>   'path': '/.Trash/Gist',
>   'description': '',
>   'uuid': '3b8d29fd-7b26-4d3e-a018-3aff68331bf2',
>   'id': 101543,
>   'doc1_id': 97429,
>   'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
>   'type': 'directory',
>   'perms': {'read': {'users': [], 'groups': []},
>    'write': {'users': [], 'groups': []},
>    'link_sharing_on': False,
>    'link_read': False,
>    'link_write': False},
>   'last_modified': '2021-02-28T13:37Z',
>   'last_modified_ts': 1614519472,
>   'is_managed': False,
>   'isSelected': False,
>   'absoluteUrl': '/home2?uuid=3b8d29fd-7b26-4d3e-a018-3aff68331bf2'},
>  {'owner': 'demo',
>   'name': 'Gist',
>   'path': '/.Trash/Gist',
>   'description': '',
>   'uuid': '1c113f15-0e21-4ef3-8279-1cd4ef96ecdd',
>   'id': 102313,
>   'doc1_id': 98169,
>   'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
>   'type': 'directory',
>   'perms': {'read': {'users': [], 'groups': []},
>    'write': {'users': [], 'groups': []},
>    'link_sharing_on': False,
>    'link_read': False,
>    'link_write': False},
>   'last_modified': '2021-02-28T13:37Z',
>   'last_modified_ts': 1614519458,
>   'is_managed': False,
>   'isSelected': False,
>   'absoluteUrl': '/home2?uuid=1c113f15-0e21-4ef3-8279-1cd4ef96ecdd'},
>  {'owner': 'demo',
>   'name': 'Gist',
>   'path': '/.Trash/Gist',
>   'description': '',
>   'uuid': '6a285c00-ec6e-4ac1-8183-04a8ba818db6',
>   'id': 108311,
>   'doc1_id': 103954,
>   'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
>   'type': 'directory',
>   'perms': {'read': {'users': [], 'groups': []},
>    'write': {'users': [], 'groups': []},
>    'link_sharing_on': False,
>    'link_read': False,
>    'link_write': False},
>   'last_modified': '2021-02-28T13:37Z',
>   'last_modified_ts': 1614519445,
>   'is_managed': False,
>   'isSelected': False,
>   'absoluteUrl': '/home2?uuid=6a285c00-ec6e-4ac1-8183-04a8ba818db6'},
>  {'owner': 'demo',
>   'name': 'Gist',
>   'path': '/.Trash/Gist',
>   'description': '',
>   'uuid': '84abbc88-74f8-4b0a-8334-43ad1b45f524',
>   'id': 108314,
>   'doc1_id': 103968,
>   'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
>   'type': 'directory',
>   'perms': {'read': {'users': [], 'groups': []},
>    'write': {'users': [], 'groups': []},
>    'link_sharing_on': False,
>    'link_read': False,
>    'link_write': False},
>   'last_modified': '2021-02-28T13:29Z',
>   'last_modified_ts': 1614518972,
>   'is_managed': False,
>   'isSelected': False,
>   'absoluteUrl': '/home2?uuid=84abbc88-74f8-4b0a-8334-43ad1b45f524'},
>  {'owner': 'demo',
>   'name': 'Gist',
>   'path': '/.Trash/Gist',
>   'description': '',
>   'uuid': '9b36c93e-4c0b-42b8-b18a-f3826dbff271',
>   'id': 108318,
>   'doc1_id': 103971,
>   'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
>   'type': 'directory',
>   'perms': {'read': {'users': [], 'groups': []},
>    'write': {'users': [], 'groups': []},
>    'link_sharing_on': False,
>    'link_read': False,
>    'link_write': False},
>   'last_modified': '2021-02-28T13:29Z',
>   'last_modified_ts': 1614518957,
>   'is_managed': False,
>   'isSelected': False,
>   'absoluteUrl': '/home2?uuid=9b36c93e-4c0b-42b8-b18a-f3826dbff271'},
>  {'owner': 'demo',
>   'name': 'Gist',
>   'path': '/.Trash/Gist',
>   'description': '',
>   'uuid': 'd24b175d-7877-4693-9e29-a749aa96216c',
>   'id': 108321,
>   'doc1_id': 103975,
>   'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
>   'type': 'directory',
>   'perms': {'read': {'users': [], 'groups': []},
>    'write': {'users': [], 'groups': []},
>    'link_sharing_on': False,
>    'link_read': False,
>    'link_write': False},
>   'last_modified': '2021-02-28T13:29Z',
>   'last_modified_ts': 1614518944,
>   'is_managed': False,
>   'isSelected': False,
>   'absoluteUrl': '/home2?uuid=d24b175d-7877-4693-9e29-a749aa96216c'},
>  {'owner': 'demo',
>   'name': 'Gist',
>   'path': '/.Trash/Gist',
>   'description': '',
>   'uuid': '206e33d2-04fd-4b68-8d3d-d5231b1b5192',
>   'id': 108326,
>   'doc1_id': 103978,
>   'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
>   'type': 'directory',
>   'perms': {'read': {'users': [], 'groups': []},
>    'write': {'users': [], 'groups': []},
>    'link_sharing_on': False,
>    'link_read': False,
>    'link_write': False},
>   'last_modified': '2021-02-28T13:28Z',
>   'last_modified_ts': 1614518915,
>   'is_managed': False,
>   'isSelected': False,
>   'absoluteUrl': '/home2?uuid=206e33d2-04fd-4b68-8d3d-d5231b1b5192'},
>  {'owner': 'demo',
>   'name': 'Gist',
>   'path': '/.Trash/Gist',
>   'description': '',
>   'uuid': 'b6807cda-9601-45c6-b43f-cb811dc6ca6c',
>   'id': 108329,
>   'doc1_id': 103983,
>   'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
>   'type': 'directory',
>   'perms': {'read': {'users': [], 'groups': []},
>    'write': {'users': [], 'groups': []},
>    'link_sharing_on': False,
>    'link_read': False,
>    'link_write': False},
>   'last_modified': '2021-02-28T13:28Z',
>   'last_modified_ts': 1614518898,
>   'is_managed': False,
>   'isSelected': False,
>   'absoluteUrl': '/home2?uuid=b6807cda-9601-45c6-b43f-cb811dc6ca6c'},
>  {'owner': 'demo',
>   'name': 'Gist',
>   'path': '/.Trash/Gist',
>   'description': '',
>   'uuid': '444727f8-5446-44cb-b554-2f759cd9092f',
>   'id': 108332,
>   'doc1_id': 103986,
>   'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
>   'type': 'directory',
>   'perms': {'read': {'users': [], 'groups': []},
>    'write': {'users': [], 'groups': []},
>    'link_sharing_on': False,
>    'link_read': False,
>    'link_write': False},
>   'last_modified': '2021-02-28T13:28Z',
>   'last_modified_ts': 1614518884,
>   'is_managed': False,
>   'isSelected': False,
>   'absoluteUrl': '/home2?uuid=444727f8-5446-44cb-b554-2f759cd9092f'},
>  {'owner': 'demo',
>   'name': 'Gist',
>   'path': '/.Trash/Gist',
>   'description': '',
>   'uuid': '4c85cf1f-8857-4f54-bbba-ede6f6ce5a12',
>   'id': 108337,
>   'doc1_id': 103989,
>   'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
>   'type': 'directory',
>   'perms': {'read': {'users': [], 'groups': []},
>    'write': {'users': [], 'groups': []},
>    'link_sharing_on': False,
>    'link_read': False,
>    'link_write': False},
>   'last_modified': '2021-02-28T11:34Z',
>   'last_modified_ts': 1614512080,
>   'is_managed': False,
>   'isSelected': False,
>   'absoluteUrl': '/home2?uuid=4c85cf1f-8857-4f54-bbba-ede6f6ce5a12'},
>  {'owner': 'demo',
>   'name': 'Gist',
>   'path': '/.Trash/Gist',
>   'description': '',
>   'uuid': '869eb70a-964a-4677-a6cd-7c1ef45566db',
>   'id': 108340,
>   'doc1_id': 103994,
>   'parent_uuid': 'bdaed5ed-926e-45e5-8754-cd5676e7180a',
>   'type': 'directory',
>   'perms': {'read': {'users': [], 'groups': []},
>    'write': {'users': [], 'groups': []},
>    'link_sharing_on': False,
>    'link_read': False,
>    'link_write': False},
>   'last_modified': '2021-02-28T11:34Z',
>   'last_modified_ts': 1614512064,
>   'is_managed': False,
>   'isSelected': False,
>   'absoluteUrl': '/home2?uuid=869eb70a-964a-4677-a6cd-7c1ef45566db'},
>  {'owner': 'demo',
>   'name': 'Gist',
>   'path': '/Gist',
>   'description': '',
>   'uuid': 'dcec6b13-1317-4a3b-98b1-621ac29542b4',
>   'id': 92640,
>   'doc1_id': 90136,
>   'parent_uuid': 'e214bac4-e46b-4fce-93f6-199c10d89a3c',
>   'type': 'directory',
>   'perms': {'read': {'users': [], 'groups': []},
>    'write': {'users': [], 'groups': []},
>    'link_sharing_on': False,
>    'link_read': False,
>    'link_write': False},
>   'last_modified': '2021-01-20T18:02Z',
>   'last_modified_ts': 1611165721,
>   'is_managed': False,
>   'isSelected': False,
>   'absoluteUrl': '/home2?uuid=dcec6b13-1317-4a3b-98b1-621ac29542b4'}]
> ```


- Reproduced the error by creating the Gist directory manually in local setup.
- Then, this patch was tested with 2 manual test scenarios --
- When dir already present and trying to create it again and when not present and then trying to create
- In demo.gethue, there were 13 Gist directories out of which I 'trashed' 12 dir ( which is indicated by 'path' ) and only 1 original one remains now which had all the gists .

So next to iterate is making a UT for this?? @romainr

**Pull request closed without merging**